Remove extra search API call

oisincoveney opened the pull request:
Closes https://github.com/metabase/metabase/issues/34001

### Description

Removes extraneous initial call to the search API when we load the `SearchResults` component.

### How to verify

1. Go to the search bar and type something quickly (for example, 'Products')
2. Previously, two calls would have been made: one for `/search?q=P` and `/search?q=Products`. Now, go to the Network tab to verify that there is only one call made

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

Commit:
Remove extra call and add test condition for calling the search only one time

oisincoveney commented on e2e/test/scenarios/onboarding/search/search.cy.spec.js:
```diff
@@ -129,6 +131,8 @@ describe("scenarios > search", () => {
         `/question/${ORDERS_QUESTION_ID}-orders`,
       );
+
+      cy.get("@search.all").should("have.length", 1);
```
Should we add this to the other tests?

nemanjaglumac replied:
I think it's safe enough to keep it only here.
It's great you added it, btw. We'll have an explicit fail mechanism in the future thanks to this.

Commit:
Add loading state for debouncing text

Commit:
uncomment reload: true

oisincoveney commented on frontend/src/metabase/nav/components/SearchResults.tsx:
```diff
@@ -34,7 +34,8 @@ export const SearchResults = ({
   const dispatch = useDispatch();
 
-  const [debouncedSearchText, setDebouncedSearchText] = useState(searchText);
+  const [debouncedSearchText, setDebouncedSearchText] = useState<string>();
```
Before, when `SearchResults` was rendered, it was calling the `useSearchListQuery` on the initial string and causing that extra call. Now, we just let the `useDebounce` hook trigger when the user types something, which will then force the `query` to only change (and cause the `useSearchListQuery` hook) to trigger only when

Commit:
Use the debouncedSearchText element for enabling search

oisincoveney commented on frontend/src/metabase/nav/components/SearchResults.tsx:
```diff
@@ -44,19 +45,17 @@ export const SearchResults = ({
   const { data: list = [], isLoading } = useSearchListQuery({
     query,
     reload: true,
+    enabled: !!debouncedSearchText,
```
We don't want to run a search if we don't have text (without this, we trigger a `/search?limit=50` query

oisincoveney commented on frontend/src/metabase/nav/components/SearchResults.tsx:
```diff
@@ -76,7 +75,7 @@ export const SearchResults = ({
   const hasResults = list.length > 0;
 
-  if (isLoading) {
+  if (isLoading || isWaitingForDebounce) {
```
Keeps the loading spinner around when the user is typing, so we're not showing old results when the user is typing

oisincoveney commented on frontend/src/metabase/nav/components/SearchResults.tsx:
```diff
@@ -44,19 +45,17 @@ export const SearchResults = ({
-    }),
-    [debouncedSearchText, searchFilters, models],
-  );
+  const query = {
```
The `useMemo` is unnecessary since the hook uses `useDeepCompareEffect` to check if `query` has changed

nemanjaglumac reviewed (approved):
Works as advertised! 🎉 

Not sure if you want to wait for FE folks to review the code changes. I'm vouching for the functionality.

oisincoveney commented:
> do we really need to backport this?

Ah. I misread the thread - yeah we shouldn't backport

Commit:
Fix unit test

kamilmielnik reviewed (changes requested):
The test does not reproduce the issue

oisincoveney commented on frontend/src/metabase/nav/components/SearchResults.tsx:
```diff
@@ -76,10 +75,10 @@ export const SearchResults = ({
     return (
       <Stack p="xl" align="center">
-        <Loader size="lg" />
+        <Loader size="lg" data-testid="loading-spinner" />
```
It is needed, there was a `waitFor` loading failure in one of the `SearchApp` unit tests now that the logic actually works as expected

oisincoveney commented on e2e/test/scenarios/onboarding/search/search.cy.spec.js:
```diff
@@ -129,6 +131,8 @@ describe("scenarios > search", () => {
         `/question/${ORDERS_QUESTION_ID}-orders`,
       );
+
+      cy.get("@search.all").should("have.length", 1);
```
It's a flaky test so it's not going to mess up every time. Try and put `cy.wait(1000)` after line 119, there should be two requests:

* `GET /search?q=o`
* `GET /search?q=ord`

Sometimes this will happen if the test runner is slow but when testing on local it's something that you won't see

kamilmielnik replied:
This change fixed it :tada: 056d97a4c4bc91d4ed1af25799b7a2fe7a424d7d

nemanjaglumac committed:
Make test more resilient

This change made sure that the test always fails on a branch without a fix.
Before this, it was sometimes giving us a false positive result.
